lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraph.java

```diff
@@ -0,0 +1,235 @@
+    bound.bound = results.top().score;
+    while (candidates.size() > 0) {
+      // get the best candidate (closest or best scoring)
+      Neighbor c = candidates.pollLast();
+      if (results.size() >= topK) {
+        if (bound.check(c.score)) {
+          break;
+        }
+      }
+      graphValues.seek(c.node);
+      int friendOrd;
+      while ((friendOrd = graphValues.nextArc()) != NO_MORE_DOCS) {
+        if (visited.contains(friendOrd)) {
+          continue;
+        }
+        visited.add(friendOrd);
+        float score = compare(query, vectors.vectorValue(friendOrd), searchStrategy);
+        if (results.size() < topK || bound.check(score) == false) {
+          Neighbor n = new Neighbor(friendOrd, score);
+          candidates.add(n);
+          results.insertWithOverflow(n);
+          bound.bound = results.top().score;
+        }
+      }
+    }
+    return results;
+  }
+
+  /**
+   * Returns the nodes connected to the given node by its outgoing arcs.
+   * @param node the node whose friends are returned
+   */
+  public int[] getNeighbors(int node) {
+    return graph.get(node).stream().mapToInt(Neighbor::node).toArray();
+  }
+
+  /** Connects two nodes symmetrically.
+   * node1 must be less than node2 and must already have been inserted to the graph */
+  void connectNodes(int node1, int node2, float score, int maxConnections) {
+    assert node1 >= 0 && node2 >= 0;
+    assert node1 < node2;
+    List<Neighbor> arcs1 = graph.get(node1);
+    assert arcs1 != null;
+    assert arcs1.isEmpty() || arcs1.get(arcs1.size() - 1).node < node2;
+    arcs1.add(new Neighbor(node2, score));
+    List<Neighbor> arcs2;
+    if (node2 < graph.size()) {
+      arcs2 = graph.get(node2);
+      assert arcs2.get(arcs2.size() - 1).node < node1;
+    } else {
+      assert node2 == graph.size();
+      arcs2 = new ArrayList<>();
+      graph.add(arcs2);
+    }
+    arcs2.add(new Neighbor(node1, score));
+
+    // ensure #arcs <= maxConnections
+    /*
+    if (maxConnections > 0) {
+      shrink(node2, maxConnections);
+    }
+     */
+  }
+
+  /**
+   * Calculates a similarity score between the two vectors with a specified function.
+   * @param v1 a vector
+   * @param v2 another vector, of the same dimension
+   * @return the value of the strategy's score function applied to the two vectors
+   */
+  public static float compare(float[] v1, float[] v2, VectorValues.SearchStrategy searchStrategy) {
+    switch (searchStrategy) {
+      case EUCLIDEAN_HNSW:
+        return squareDistance(v1, v2);
+      case DOT_PRODUCT_HNSW:
+        return dotProduct(v1, v2);
+      default:
+        throw new IllegalStateException("invalid search strategy: " + searchStrategy);
+    }
+  }
+
+  /**
+   * Return whether the given strategy returns lower values for nearer vectors
+   * @param searchStrategy the strategy
+   */
+  public static boolean isReversed(VectorValues.SearchStrategy searchStrategy) {
```

Comment:
Maybe we could add this as a method on the `enum` constants instead?  So then caller could just call `SearchStrategy.isReversed`?

Reply:
yeah that does seem cleaner. It was what I had in an earlier revision, but got thrown out with the bathwater of the vector scoring implementations.